python/cudf/cudf/core/buffer/tenable_buffer.py

```diff
@@ -0,0 +1,302 @@
+        return BufferSlice(
+            base=self._base, offset=offset + self._offset, size=size
+        )
+
+    @property
+    def exposed(self) -> bool:
+        return self._base.exposed
+
+    def get_ptr(self, mode: str = "write") -> int:
+        if mode == "write" and cudf.get_option("copy_on_write"):
+            self.make_single_owner_inplace()
+        return self._base.get_ptr(mode=mode) + self._offset
+
+    def memoryview(
+        self, *, offset: int = 0, size: Optional[int] = None
+    ) -> memoryview:
+        return self._base.memoryview(offset=self._offset + offset, size=size)
+
+    def copy(self, deep: bool = True) -> BufferSlice:
+        """Return a copy of Buffer.
+
+        What actually happens when `deep == False` depends on the
+        "copy_on_write" option. When copy-on-write is enabled, a shallow copy
+        because a deep copy if the buffer has been exposed. This is because we
+        have no control over knowing if the data is being modified when the
+        buffer has been exposed to third-party.
+
+        Parameters
+        ----------
+        deep : bool, default True
+            The semantics when copy-on-write is disabled:
+                - If True, returns a deep copy of the underlying Buffer data.
+                - If False, returns a shallow copy of the Buffer pointing to
+                  the same underlying data.
+            The semantics when copy-on-write is enabled:
+                - Always a deep copy of the underlying Buffer data.
+
+        Returns
+        -------
+        BufferSlice
+            A slice pointing to either a new or the existing base buffer
+            depending on the expose status of the base buffer and the
+            copy-on-write option (see above).
+        """
+        if deep or not cudf.get_option("copy_on_write"):
+            base_copy = self._base.copy(deep=deep)
+        else:
+            base_copy = self._base.copy(deep=self.exposed)
+        return cast(
+            BufferSlice, base_copy[self._offset : self._offset + self._size]
+        )
+
+    @property
+    def __cuda_array_interface__(self) -> Mapping:
+        if cudf.get_option("copy_on_write"):
+            self.make_single_owner_inplace()
+        return super().__cuda_array_interface__
+
+    def make_single_owner_inplace(self) -> None:
+        """Make sure this slice is the only one pointing to the base.
+
+        This is used by copy-on-write to trigger a deep copy when write
+        access is detected.
+
+        Parameters
+        ----------
+        data : device-buffer-like
+            An object implementing the CUDA Array Interface.
+
+        Returns
+        -------
+        Buffer
+            Buffer representing the same device memory as `data`
+        """
+
+        if len(self._base._slices) > 1:
+            # If this is not the only slice pointing to `self._base`, we
+            # point to a new deep copy of the base.
+            t = self.copy(deep=True)
```

Comment:
Why do we not just copy `self._base` directly?

Reply:
This is to support `SpillableBuffer` when it also uses `BufferSlice`